### ISSUE TITLE
Can tau_rand return values < 0 ?

### ISSUE THREAD

**Issue by jlmelville** [lmcinnes/umap#202]
The description of `tau_rand` says that it returns "A (pseudo)-random float32 in the interval [0, 1]". Using 
version 0.3.7 (installed via Anaconda on Windows 10), it seems to generate floats in the interval [-1, 1]:

```python
import umap
import numpy as np

rng_state = np.array([1000, 2000, 3000], dtype = np.int64)
for i in range(10):
     print(umap.utils.tau_rand(rng_state))

0.18258631229400635
-0.18693089485168457
0.1864445060491562
0.3614959716796875
-0.8346158862113953
-0.5668963193893433
-0.7467218637466431
0.7167382836341858
-0.21391557157039642
-0.009349984116852283
```

If I am doing it right, this would seem to have implications only for the nearest neighbor descent part of the code, but I think it means that the sample rate, `rho` is effectively larger than expected. 

In the current master, it effects the main loop of the `nn_descent` code. I see there have been some changes on the 0.4 branch, where I think it would have an effect on `new_build_candidates`?

**Comment by lmcinnes** [comment]
Yes, that looks suspiciously bad :disappointed: 

I'll have to look at it soon, but I suspect it is a simple error that has gone undetected because it had limited overall impact on results. Ideally this will actually make everything better and faster. I can hope...

**Comment by jlmelville** [comment]
I think the problem is to do with `tau_rand_int` returning signed rather than unsigned ints. Using the bitmask to make it signed seems to work, e.g.:

```python
return (integer & 0x7fffffff) / 0x7fffffff
```

**Comment by lmcinnes** [comment]
Yes, that is exactly the problem. Oddly enough in benchmarking just using ``abs`` in the code is as quick as the bitwise and, and is probably more clear, so I'll go with that. Thanks for catching this! I believe this has probably been an issue for a long time.

**Comment by jlmelville** [comment]
Confirmed fixed in 0.3.8.
